app/controllers/submissions_controller.rb

```diff
@@ -198,6 +236,51 @@ def downloadAll
               filename: "#{@course.name}_#{@course.semester}_#{@assessment.name}_submissions.zip")
   end
 
+  action_auth_level :downloadBatch, :course_assistant
+  def downloadBatch
+    submission_ids = params[:submission_ids]
+    flash[:error] = "Cannot index submissions for nil assessment" if @assessment.nil?
+
+    unless @assessment.valid?
+      @assessment.errors.full_messages.each do |msg|
+        flash[:error] += "<br>#{msg}"
+      end
+      flash[:html_safe] = true
+    end
+
+    if @assessment.disable_handins
+      flash[:error] = "There are no submissions to download."
+      if @cud.course_assistant
+        redirect_to([@course, @assessment])
+      else
+        redirect_to([@course, @assessment, :submissions])
+      end
+      return
+    end
+
+    submissions = submission_ids.map { |sid| @assessment.submissions.find_by(sid) }
+    submissions = submissions.select { |s| @cud.can_administer?(s.course_user_datum) }
+    paths = submissions.collect(&:handin_file_path)
+    paths = paths.select { |p| !p.nil? && File.exist?(p) && File.readable?(p) }
+
+    result = Archive.create_zip paths # result is stringIO to be sent
+
+    if result.nil?
```

Comment:
nit: since you're using this error flash & redirect code more than once, it might be good to write a helper method to do all this in order to make things more modular

Reply:
Fixed